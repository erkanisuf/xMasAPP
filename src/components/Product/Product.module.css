.productContainer {
  display: flex;
  width: 90%;
  flex: 1;
  margin: 0 auto;
  padding: 5px;
  justify-content: space-evenly;
  align-items: flex-start;

  max-height: 150px;
  padding-bottom: 15px;
}
.productContainer img {
  width: 100px;
  height: 100px;
  object-fit: contain;
  border: 1px solid #ccc;
  padding: 15px;
  margin: 5px;
}
.productContainer p {
  font-size: 12px;
}
.productContainer span {
  font-size: 22px;
  font-weight: 700;
}
