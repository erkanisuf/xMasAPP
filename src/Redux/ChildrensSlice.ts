import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Product } from "../components/ChildComp/ChildComp";
import { IProduct } from "../components/Product/Product";
import type { RootState } from "./store";

interface IReduxChildrenSlice {
  ChildrenApprovedItems: OBJtoAPI[];
  ChildrenDiscardedItems: OBJtoAPI[];
}
export interface OBJtoAPI {
  userId: number;
  date: Date | string;
  products: Product[];
}
// This is the Initial State represents the Products from the Open APi
const initialState: IReduxChildrenSlice = {
  ChildrenApprovedItems: [],
  ChildrenDiscardedItems: [],
};

export const ChildrensSlice = createSlice({
  name: "childrenSlice",
  //Redux Info
  // `createSlice` will infer the state type from the `initialState` argument
  initialState,
  reducers: {
    //Redux Tips
    // Use the PayloadAction type to declare the contents of `action.payload`
    ChildProductToApproved: (state, action: PayloadAction<OBJtoAPI>) => {
      const copyApprovedState = [...state.ChildrenApprovedItems];
      const findIndex = copyApprovedState.findIndex(
        (el) => el.userId === action.payload.userId
      );
      //If the user is already there , it finds its index and just updates the products array.
      if (findIndex !== -1) {
        copyApprovedState[findIndex].products.push(action.payload.products[0]);
      } else {
        copyApprovedState.push(action.payload);
      }

      state.ChildrenApprovedItems = copyApprovedState;
    },
    ChildProductToDiscarded: (state, action: PayloadAction<OBJtoAPI>) => {
      const copyDiscardedState = [...state.ChildrenDiscardedItems];
      const findIndex = copyDiscardedState.findIndex(
        (el) => el.userId === action.payload.userId
      );
      //If the user is already there , it finds its index and just updates the products array.
      if (findIndex != -1) {
        copyDiscardedState[findIndex].products.push(action.payload.products[0]);
      } else {
        copyDiscardedState.push(action.payload);
      }

      state.ChildrenDiscardedItems = copyDiscardedState;
    },
  },
});

export const {
  ChildProductToApproved,
  ChildProductToDiscarded,
} = ChildrensSlice.actions;

//Redux Setup from their Docs.
// Other code such as selectors can use the imported `RootState` type
export const selectCount = (state: RootState) => state;

export default ChildrensSlice.reducer;
