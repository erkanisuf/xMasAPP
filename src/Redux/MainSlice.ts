import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { IProductDesc } from "../components/Product/Product";
import type { RootState } from "./store";

interface IReduxMainSlice {
  products: IProductDesc[];
  myChildren: { id: number; name: string; image: string }[];
}
interface IChangePrice {
  id: number;
  discount: number;
}

// This is the Initial State represents the Products from the Open APi
const initialState: IReduxMainSlice = {
  products: [],
  myChildren: [
    {
      id: 1,
      name: "Pekka",
      image:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSVY3efFyktSBShjhBcNUt2JuH3F1r0qnlt1g&usqp=CAU",
    },
    {
      id: 2,
      name: "Mikko",
      image:
        "https://www.pngitem.com/pimgs/m/424-4240458_one-ok-rock-stand-out-fit-in-from.png",
    },
    {
      id: 3,
      name: "Juha",
      image:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRdBkq9ItXGCTqTIW4BdBlaEpZQVllKv8gkQGpdhVwfbAIuj9-4yOzapTjdNE0r14wSJoE&usqp=CAU",
    },
    {
      id: 4,
      name: "Jonna",
      image:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQlCT3cJxdz3-ntu8ixFRexArdaknweGrBQrw&usqp=CAU",
    },
    {
      id: 5,
      name: "Erkku",
      image:
        "https://st4.depositphotos.com/18690434/20926/v/450/depositphotos_209266078-stock-illustration-boy-kid-avatar-icon-vector.jpg",
    },
  ],
};

export const MainSlice = createSlice({
  name: "main",
  //Redux Info
  // `createSlice` will infer the state type from the `initialState` argument
  initialState,
  reducers: {
    //Redux Tips
    // Use the PayloadAction type to declare the contents of `action.payload`
    addProductsFromApi: (state, action: PayloadAction<IProductDesc[]>) => {
      state.products = action.payload;
    },
    changeProductPrice: (state, action: PayloadAction<IChangePrice>) => {
      const copyProducts = [...state.products];
      const findIndex = copyProducts.findIndex(
        (el) => el.id === action.payload.id
      );
      //Discount if more than 1 quantity
      if (findIndex !== -1) {
        copyProducts[findIndex] = {
          ...copyProducts[findIndex],
          discount: action.payload.discount,
        };
      }
      state.products = copyProducts;
    },
  },
});

export const { addProductsFromApi, changeProductPrice } = MainSlice.actions;

//Redux Setup from their Docs.
// Other code such as selectors can use the imported `RootState` type
export const selectCount = (state: RootState) => state;

export default MainSlice.reducer;
