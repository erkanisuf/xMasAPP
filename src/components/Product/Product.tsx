import React, { useEffect, useState } from "react";
import { useAppSelector } from "../../Redux/hooks";
import ProductCSS from "./Product.module.css"; // Styling

export interface IProduct {
  id: number;
  title: string;
  price: number;
  description: string;
  category: string;
  image: string;
  discount?: number;
}

export interface IProductProp {
  productId: number;
}

export interface IuseFetch {
  response: IProduct;
  isLoading: boolean;
  error: any;
}
//Products are searched from the redux main state with .fid(), another alternative would be to fetch single product from the API but it feels slower.
const Product: React.FC<IProductProp> = ({ productId }) => {
  const allMainStateProducts = useAppSelector((state) => state.main.products); // Redux Approved Childen list
  const reduxState = useAppSelector((state) => state.main); // Redux Hook of State
  const reduxStateProducts = useAppSelector((state) => state.childrens); // Redux Hook of State
  const [product, setProduct] = useState<IProduct>();

  useEffect(() => {
    //Finds the product from the State
    const findItemFromState = () => {
      const findProduct = reduxState.products.find((el) => el.id === productId);
      setProduct(findProduct);
    };
    findItemFromState();

    return () => {};
  }, [
    productId,
    reduxState.products,
    reduxStateProducts,
    allMainStateProducts,
  ]);

  return (
    <div className={ProductCSS.productContainer}>
      <div>
        <img src={product?.image} alt={product?.title} />
      </div>
      <div>
        <div>
          <p>{product?.title}</p>
        </div>
        <div>
          {" "}
          <span>{product?.price.toFixed(2)} â‚¬</span>
          {product?.discount ? (
            <span style={{ color: "red" }}>{product.discount}%</span>
          ) : (
            ""
          )}
        </div>
      </div>
    </div>
  );
};

export default React.memo(Product);
