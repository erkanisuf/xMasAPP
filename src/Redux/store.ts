import { configureStore } from "@reduxjs/toolkit";
import ChildrensSlice from "./ChildrensSlice";
import MainSlice from "./MainSlice";
// ...

export const store = configureStore({
  reducer: { main: MainSlice, childrens: ChildrensSlice },
});

// Redux required setup from Their Docs
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
