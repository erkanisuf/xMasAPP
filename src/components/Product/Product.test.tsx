import React from "react";
import renderer from "react-test-renderer";

import ReactDOM from "react-dom";
import { render } from "@testing-library/react";

import Product from "./Product";
import { Provider } from "react-redux";
import { store } from "../../Redux/store";
test("Fake", () => {
  expect(true).toBeTruthy();
});

it("red", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Product productId={1} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
