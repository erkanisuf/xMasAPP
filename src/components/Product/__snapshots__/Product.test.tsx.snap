// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`red 1`] = `
<div
  className="productContainer"
>
  <div>
    <img />
  </div>
  <div>
    <div>
      <p />
    </div>
    <div>
       
      <span>
         â‚¬
      </span>
    </div>
  </div>
</div>
`;
